File: project-{topic}-pro/backend/src/controllers/project.controller.ts
```typescript
import { Request, Response, NextFunction } from 'express';
import projectService from '@services/project.service';
import logger from '@config/logger';

class ProjectController {
  async createProject(req: Request, res: Response, next: NextFunction) {
    try {
      const projectData = req.body;
      const ownerId = req.user!.id; // Authenticated user is the owner
      const project = await projectService.createProject({ ...projectData, ownerId });
      res.status(201).json(project);
    } catch (error) {
      logger.error(`Create project controller error: ${error instanceof Error ? error.message : error}`);
      next(error);
    }
  }

  async getAllProjects(req: Request, res: Response, next: NextFunction) {
    try {
      const projects = await projectService.getAllProjects(); // Admin can see all
      res.status(200).json(projects);
    } catch (error) {
      logger.error(`Get all projects controller error: ${error instanceof Error ? error.message : error}`);
      next(error);
    }
  }

  async getProjectById(req: Request, res: Response, next: NextFunction) {
    try {
      const { id } = req.params;
      const project = await projectService.getProjectById(id);
      res.status(200).json(project);
    } catch (error) {
      logger.error(`Get project by ID controller error: ${error instanceof Error ? error.message : error}`);
      next(error);
    }
  }

  async updateProject(req: Request, res: Response, next: NextFunction) {
    try {
      const { id } = req.params;
      const updateData = req.body;
      const userId = req.user!.id;
      const userRole = req.user!.role;
      const updatedProject = await projectService.updateProject(id, updateData, userId, userRole);
      res.status(200).json(updatedProject);
    } catch (error) {
      logger.error(`Update project controller error: ${error instanceof Error ? error.message : error}`);
      next(error);
    }
  }

  async deleteProject(req: Request, res: Response, next: NextFunction) {
    try {
      const { id } = req.params;
      const userId = req.user!.id;
      const userRole = req.user!.role;
      await projectService.deleteProject(id, userId, userRole);
      res.status(204).send();
    } catch (error) {
      logger.error(`Delete project controller error: ${error instanceof Error ? error.message : error}`);
      next(error);
    }
  }
}

export default new ProjectController();
```

File: project-{topic}-advanced/src/controllers/ProjectController.cpp
```cpp
#include "ProjectController.hpp"

ProjectController::ProjectController(std::shared_ptr<ProjectService> project_service,
                                     std::shared_ptr<UserService> user_service)
    : project_service_(project_service), user_service_(user_service) {}

void ProjectController::registerRoutes(crow::App<crow::CORSHandler, AuthMiddleware, RateLimitMiddleware>& app) {
    // Get all projects
    CROW_ROUTE(app, "/api/v1/projects")
        .methods(crow::HTTPMethod::GET)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx) {
        return try_catch_handler([&]() -> crow::response {
            std::vector<Project> projects = project_service_->getAllProjects();
            return crow::response(200, JSONConverter::toJSON(projects).dump());
        });
    });

    // Create a new project
    CROW_ROUTE(app, "/api/v1/projects")
        .methods(crow::HTTPMethod::POST)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx) {
        return try_catch_handler([&]() -> crow::response {
            nlohmann::json req_body = JSONConverter::parse(req.body);
            ProjectCreateDTO create_dto = ProjectCreateDTO::fromJson(req_body);

            // Authorization: If owner_id is specified and it's not the authenticated user,
            // only admins can set a different owner.
            if (create_dto.owner_id.has_value() && create_dto.owner_id.value() != ctx.user_id) {
                if (!AuthMiddleware::hasRole(ctx, UserRole::ADMIN)) {
                    throw ForbiddenException("Access denied: Only administrators can create projects for other users.");
                }
                // Also validate that the specified owner_id actually exists
                if (!user_service_->getUserById(create_dto.owner_id.value()).has_value()) {
                    throw BadRequestException("Owner user with ID " + std::to_string(create_dto.owner_id.value()) + " does not exist.");
                }
            }
            
            // If owner_id not provided, the current user is the owner
            if (!create_dto.owner_id.has_value()) {
                create_dto.owner_id = ctx.user_id;
            }

            Project new_project = project_service_->createProject(create_dto, ctx.user_id); // Pass auth user ID for implicit ownership
            return crow::response(201, JSONConverter::toJSON(new_project).dump());
        });
    });

    // Get project by ID
    CROW_ROUTE(app, "/api/v1/projects/<int>")
        .methods(crow::HTTPMethod::GET)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int project_id) {
        return try_catch_handler([&]() -> crow::response {
            std::optional<Project> project = project_service_->getProjectById(project_id);
            if (!project.has_value()) {
                throw NotFoundException("Project with ID " + std::to_string(project_id) + " not found.");
            }
            // Authorization: Only owner or admin can view project details (or public projects if applicable)
            // For now, assume all projects are viewable by any authenticated user for simplicity,
            // but add a check for modification/deletion.
            return crow::response(200, JSONConverter::toJSON(project.value()).dump());
        });
    });

    // Update project by ID
    CROW_ROUTE(app, "/api/v1/projects/<int>")
        .methods(crow::HTTPMethod::PUT)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int project_id) {
        return try_catch_handler([&]() -> crow::response {
            // Authorization: Only project owner or admin can update
            if (!AuthMiddleware::hasRole(ctx, UserRole::ADMIN) && !project_service_->isProjectOwner(project_id, ctx.user_id)) {
                throw ForbiddenException("Access denied: Not authorized to update this project.");
            }

            nlohmann::json req_body = JSONConverter::parse(req.body);
            ProjectUpdateDTO update_dto = ProjectUpdateDTO::fromJson(req_body);

            // If `owner_id` is updated, ensure only admin can do it, and the new owner exists.
            if (update_dto.owner_id.has_value()) {
                if (!AuthMiddleware::hasRole(ctx, UserRole::ADMIN)) {
                    throw ForbiddenException("Access denied: Only administrators can change project ownership.");
                }
                if (update_dto.owner_id.value() != 0 && !user_service_->getUserById(update_dto.owner_id.value()).has_value()) {
                    throw BadRequestException("New owner user with ID " + std::to_string(update_dto.owner_id.value()) + " does not exist.");
                }
            } else if (req_body.contains("owner_id") && req_body["owner_id"].is_null()) { // Explicit null
                if (!AuthMiddleware::hasRole(ctx, UserRole::ADMIN)) {
                    throw ForbiddenException("Access denied: Only administrators can remove project ownership.");
                }
            }

            Project updated_project = project_service_->updateProject(project_id, update_dto);
            return crow::response(200, JSONConverter::toJSON(updated_project).dump());
        });
    });

    // Delete project by ID
    CROW_ROUTE(app, "/api/v1/projects/<int>")
        .methods(crow::HTTPMethod::DELETE)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int project_id) {
        return try_catch_handler([&]() -> crow::response {
            // Authorization: Only project owner or admin can delete
            if (!AuthMiddleware::hasRole(ctx, UserRole::ADMIN) && !project_service_->isProjectOwner(project_id, ctx.user_id)) {
                throw ForbiddenException("Access denied: Not authorized to delete this project.");
            }

            project_service_->deleteProject(project_id);
            return crow::response(204); // No content
        });
    });
}
```